src/java.base/macosx/classes/sun/nio/ch/KQueue.java

```diff
@@ -47,74 +60,113 @@ private KQueue() { }
      *        void            *udata;         // opaque user data identifier
      * };
      */
-    static {
-        IOUtil.load();
-    }
-    private static final int SIZEOF_KQUEUEEVENT    = keventSize();
-    private static final int OFFSET_IDENT          = identOffset();
-    private static final int OFFSET_FILTER         = filterOffset();
-    private static final int OFFSET_FLAGS          = flagsOffset();
 
     // filters
-    static final int EVFILT_READ  = -1;
-    static final int EVFILT_WRITE = -2;
+    static final int EVFILT_READ  = kqueue_h.EVFILT_READ();
+    static final int EVFILT_WRITE = kqueue_h.EVFILT_WRITE();
 
     // flags
-    static final int EV_ADD     = 0x0001;
-    static final int EV_DELETE  = 0x0002;
-    static final int EV_ONESHOT = 0x0010;
-    static final int EV_CLEAR   = 0x0020;
+    static final int EV_ADD     = kqueue_h.EV_ADD();
+    static final int EV_DELETE  = kqueue_h.EV_DELETE();
+    static final int EV_ONESHOT = kqueue_h.EV_ONESHOT();
+    static final int EV_CLEAR   = kqueue_h.EV_CLEAR();
 
     /**
      * Allocates a poll array to handle up to {@code count} events.
      */
-    static long allocatePollArray(int count) {
-        return unsafe.allocateMemory(count * SIZEOF_KQUEUEEVENT);
+    static MemorySegment allocatePollArray(int count) {
+        return kevent.allocateArray(count, FFMUtils.SEGMENT_ALLOCATOR);
     }
 
     /**
      * Free a poll array
      */
-    static void freePollArray(long address) {
-        unsafe.freeMemory(address);
+    static void freePollArray(MemorySegment memorySegment){
+        FFMUtils.free(memorySegment);
     }
 
     /**
      * Returns kevent[i].
      */
-    static long getEvent(long address, int i) {
-        return address + (SIZEOF_KQUEUEEVENT*i);
+    static MemorySegment getEvent(MemorySegment memoryHandle, int i) {
+        return kevent.asSlice(memoryHandle, i);
     }
 
     /**
      * Returns the file descriptor from a kevent (assuming it is in the ident field)
      */
-    static int getDescriptor(long address) {
-        return unsafe.getInt(address + OFFSET_IDENT);
+    static long getDescriptor(MemorySegment memoryHandle) {
+        return kevent.ident(memoryHandle);
```

Comment:
Do you mind going back over the discussion we had in the original PR? The ident field is a 64-bit, an fd is an int. This means that KQueue should have a getDescriptor that returns an int, or a getIdent that returns a long. The former is cleaner at the use-sites.